cpp/src/binaryop/compiled/util.cpp

```diff
@@ -0,0 +1,122 @@
+        case binary_operator::DIV:                  return call<ops::Div, TypeOut>();
+        case binary_operator::TRUE_DIV:             return call<ops::TrueDiv, TypeOut>();
+        case binary_operator::FLOOR_DIV:            return call<ops::FloorDiv, TypeOut>();
+        case binary_operator::MOD:                  return call<ops::Mod, TypeOut>();
+        case binary_operator::PYMOD:                return call<ops::PyMod, TypeOut>();
+        case binary_operator::POW:                  return call<ops::Pow, TypeOut>();
+        case binary_operator::EQUAL:                return call<ops::Equal, TypeOut>();
+        case binary_operator::NOT_EQUAL:            return call<ops::NotEqual, TypeOut>();
+        case binary_operator::LESS:                 return call<ops::Less, TypeOut>();
+        case binary_operator::GREATER:              return call<ops::Greater, TypeOut>();
+        case binary_operator::LESS_EQUAL:           return call<ops::LessEqual, TypeOut>();
+        case binary_operator::GREATER_EQUAL:        return call<ops::GreaterEqual, TypeOut>();
+        case binary_operator::BITWISE_AND:          return call<ops::BitwiseAnd, TypeOut>();
+        case binary_operator::BITWISE_OR:           return call<ops::BitwiseOr, TypeOut>();
+        case binary_operator::BITWISE_XOR:          return call<ops::BitwiseXor, TypeOut>();
+        case binary_operator::LOGICAL_AND:          return call<ops::LogicalAnd, TypeOut>();
+        case binary_operator::LOGICAL_OR:           return call<ops::LogicalOr, TypeOut>();
+        //case binary_operator::GENERIC_BINARY:       return call<ops::UserDefinedOp, TypeOut>();
+        case binary_operator::SHIFT_LEFT:           return call<ops::ShiftLeft, TypeOut>();
+        case binary_operator::SHIFT_RIGHT:          return call<ops::ShiftRight, TypeOut>();
+        case binary_operator::SHIFT_RIGHT_UNSIGNED: return call<ops::ShiftRightUnsigned, TypeOut>();
+        case binary_operator::LOG_BASE:             return call<ops::LogBase, TypeOut>();
+        case binary_operator::ATAN2:                return call<ops::ATan2, TypeOut>();
+        case binary_operator::PMOD:                 return call<ops::PMod, TypeOut>();
+        /*
+        case binary_operator::NULL_EQUALS:          return call<ops::NullEquals, TypeOut>();
+        case binary_operator::NULL_MAX:             return call<ops::NullMax, TypeOut>();
+        case binary_operator::NULL_MIN:             return call<ops::NullMin, TypeOut>();
+        */
+        default:                                    return false;
+          // clang-format on
+      }
+    }
+  };
+  template <typename TypeLhs, typename TypeRhs>
+  inline constexpr bool operator()(data_type out, binary_operator op)
+  {
+    return type_dispatcher(out, nested_support_functor<TypeLhs, TypeRhs>{}, op);
+  }
+};
+
+}  // namespace
+
+data_type get_common_type(data_type out, data_type lhs, data_type rhs)
```

Comment:
I think it would be nicer if this returned a `std::optional<data_type>` instead of using `EMPTY` to indicate the lack of a common type.

Reply:
done. (increases compile time of util.cpp)